### PR TITLE
Start work on supporting conditionals (let's start with if-else)

### LINKED ISSUE #617: Start work on supporting conditionals (let's start with if-else)
We need basic support for if-else structures so we can show something cool at the intermediate fair.

### PR DESCRIPTION
**Pull request by LukasPietzschmann**
closes #617

### CODE REVIEW

File: src/abstract-interpretation/ainode.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
+import { NodeId, ParentInformation, RNodeWithParent } from '../r-bridge'
+import { assertUnreachable, guard } from '../util/assert'
+import { Domain, unifyDomains } from './domain'
+
+export interface AINode {
+	// The ID of the node that logically holds the domain
+	readonly nodeId:       NodeId
+	// The ID of the whole expression that the domain was calculated from
+	readonly expressionId: NodeId
```
Comment (EagleoutIce) [inline src/abstract-interpretation/ainode.ts:9]: is expression the correct reference here? With the separation being arbitrary an example may be beneficial

File: src/abstract-interpretation/ainode.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
+import { NodeId, ParentInformation, RNodeWithParent } from '../r-bridge'
+import { assertUnreachable, guard } from '../util/assert'
+import { Domain, unifyDomains } from './domain'
+
+export interface AINode {
+	// The ID of the node that logically holds the domain
+	readonly nodeId:       NodeId
```
Comment (EagleoutIce) [inline src/abstract-interpretation/ainode.ts:7]: why not join this with the respective astNode? as they are inherently linked by the traversal it could be a direct decoration.

File: src/abstract-interpretation/ainode.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+		} else if(content !== undefined) {
+			super([[content.nodeId, content]])
+		} else if(content === undefined) {
+			super()
+		} else {
+			assertUnreachable(content)
+		}
+	}
+
+	push(node: AINode): void {
```
Comment (EagleoutIce) [inline src/abstract-interpretation/ainode.ts:28]: given the map nature, something like `register` is probably a better name.

Comment (EagleoutIce) [reply src/abstract-interpretation/ainode.ts:28]: \+ given that we add a wrapper to use, how about guarding against overwriting?
I mean yes, we assume that every node is unique, but what if we add the same node multiple times.

File: src/abstract-interpretation/ainode.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+		this.set(node.nodeId, node)
+	}
+}
+
+export function mergeDomainStores(...stores: AINodeStore[]): AINodeStore {
```
Comment (EagleoutIce) [inline src/abstract-interpretation/ainode.ts:33]:
```suggestion
export function mergeDomainStores(...stores: readonly AINodeStore[]): AINodeStore {
```

File: src/abstract-interpretation/ainode.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+	const result = new AINodeStore()
+	for(const store of stores) {
+		for(const [id, node] of store) {
+			if(result.has(id)) {
+				const existing = result.get(id)
+				guard(existing !== undefined, `Domain for ID ${id} is missing`)
+				const unified = unifyDomains([existing.domain, node.domain])
+				result.set(id, {...node, domain: unified})
```
Comment (EagleoutIce) [inline src/abstract-interpretation/ainode.ts:41]: why is this not using the `push`/`register` functionality?

File: src/abstract-interpretation/ainode.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+export function mergeDomainStores(...stores: AINodeStore[]): AINodeStore {
+	const result = new AINodeStore()
+	for(const store of stores) {
+		for(const [id, node] of store) {
+			if(result.has(id)) {
+				const existing = result.get(id)
+				guard(existing !== undefined, `Domain for ID ${id} is missing`)
+				const unified = unifyDomains([existing.domain, node.domain])
```
Comment (EagleoutIce) [inline src/abstract-interpretation/ainode.ts:40]: Theoretically we can combine this as a fold/reduction providing the merging function - just in case in the future we add more than the domain.

File: src/abstract-interpretation/domain.ts [hunk 7]
```diff
@@ -90,23 +90,41 @@ function compareIntervalsByTheirMaximum(interval1: Interval, interval2: Interval
 	return compareIntervals(CompareType.Max, interval1.max, interval2.max)
 }
 
-export function doIntervalsOverlap(interval1: Interval, interval2: Interval): boolean {
+export const enum OverlapKind {
+	Overlap = 0,
+	Touch = 1,
+}
+
+export function doIntervalsOverlap(interval1: Interval, interval2: Interval, kind: OverlapKind = OverlapKind.Overlap): boolean {
```
Comment (EagleoutIce) [inline src/abstract-interpretation/domain.ts:98]: We definitely need tests for this.

File: src/abstract-interpretation/domain.ts [hunk 8]
```diff
@@ -90,23 +90,41 @@ function compareIntervalsByTheirMaximum(interval1: Interval, interval2: Interval
...
+	let doIntervalsTouch = true
+
 	// If one interval ends before the other starts, they don't overlap
 	if(diff1 < 0 || diff2 < 0) {
-		return false
-	}
-	// If their end and start are equal, they only overlap if both are inclusive
-	if(diff1 === 0) {
-		return interval1.max.inclusive && interval2.min.inclusive
+		doIntervalsOverlap = false
```
Comment (EagleoutIce) [inline src/abstract-interpretation/domain.ts:107]: It should suffice to track Overlap based on the requested mode.

File: src/abstract-interpretation/domain.ts (hunk 7 above)
Comment (EagleoutIce) [reply src/abstract-interpretation/domain.ts:98]: and fundamentally i think the function would be simpler if handled as an intersection of interval, checking if it is empty/scalar.

File: src/abstract-interpretation/domain.ts [hunk 9]
```diff
@@ -166,4 +184,82 @@ export function subtractDomains(domain1: Domain, domain2: Domain): Domain {
...
+	intersection: Interval | undefined,
+	larger:       Interval | undefined
+}
+
+function flipInclusiveness(intervalBound: IntervalBound): IntervalBound {
+	return {value: intervalBound.value, inclusive: !intervalBound.inclusive}
+
+}
+
+export function overlapIntervals(interval1: Interval, interval2: Interval): IntervalOverlap {
```
Comment (EagleoutIce) [inline src/abstract-interpretation/domain.ts:206]: warum das und `doIntervalsOverlap`?

File: src/util/cfg/cfg.ts [hunk 10]
```diff
@@ -524,24 +524,25 @@ function cfgUnaryOp(unary: RNodeWithParent, operand: ControlFlowInformation): Co
 
 
 function cfgExprList(_node: RNodeWithParent, expressions: ControlFlowInformation[]): ControlFlowInformation {
```
Comment (EagleoutIce) [inline src/util/cfg/cfg.ts:526]: not too sure whether we just just switch to basic blocks

File: src/abstract-interpretation/ainode.ts (hunk 1 above)
Comment (LukasPietzschmann) [reply src/abstract-interpretation/ainode.ts:9]: I added an example to each comment. But as we're currently not using the `expressionId` at all, I might just remove it in the future (tbh I can't really remember why I added it in the first place 😅)

Comment (EagleoutIce) [reply src/abstract-interpretation/ainode.ts:9]: I think it was to indicated the respective parent expression when operating on the sub-expression level.

File: src/abstract-interpretation/ainode.ts (hunk 3 above)
Comment (LukasPietzschmann) [reply src/abstract-interpretation/ainode.ts:28]: Sure, fixed it :) I added a `guard(...)` that guards against attempts that want to overwrite a value. When I spot a case that triggers the guard, I'll reevaluate if this is correct behavior (but it's probably not).

File: src/abstract-interpretation/ainode.ts (hunk 5 above)
Comment (LukasPietzschmann) [reply src/abstract-interpretation/ainode.ts:41]: I also just noticed this when renaming the function 😅 I probably just forgot

File: src/abstract-interpretation/domain.ts (hunk 7 above)
Comment (LukasPietzschmann) [reply src/abstract-interpretation/domain.ts:98]: > We definitely need tests for this.

We already have some (six), but they don't yet cover all edge cases.

> and fundamentally i think the function would be simpler if handled as an intersection of interval, checking if it is empty/scalar.

This would be definitly more elegant and reusable. I'll write it down somewhere and change it when I get to it. For now, I prioritize making progress; refactorings like this can come later.

File: src/abstract-interpretation/domain.ts (hunk 9 above)
Comment (LukasPietzschmann) [reply src/abstract-interpretation/domain.ts:206]: `overlapIntervals` currently uses `doIntervalsOverlap`, but this could definitly be refactored, then `doIntervalsOverlap` would be no longer needed (or maybe just as a small wrapper for `overlapIntervals`).

File: src/abstract-interpretation/ainode.ts (hunk 3 above)
Comment (LukasPietzschmann) [reply src/abstract-interpretation/ainode.ts:28]: Nvm, we do indeed use `register` to overwrite values. I added an extra flag that should be set when overwriting should be possible.

File: src/abstract-interpretation/domain.ts (hunk 7 above)
Comment (LukasPietzschmann) [reply src/abstract-interpretation/domain.ts:143]: This is now tracked in #588

File: src/abstract-interpretation/ainode.ts (hunk 2 above)
Comment (LukasPietzschmann) [reply src/abstract-interpretation/ainode.ts:7]: Addressed in e6d3add0

File: src/util/cfg/cfg.ts (hunk 10 above)
Comment (LukasPietzschmann) [reply src/util/cfg/cfg.ts:526]: Tracked in #812
